core-primitives/stf-state-handler/src/state_snapshot_repository_loader.rs

```diff
@@ -0,0 +1,220 @@
+	},
+	state_snapshot_repository::StateSnapshotRepository,
+};
+use itp_types::ShardIdentifier;
+use log::*;
+use std::{
+	collections::VecDeque, fmt::Debug, iter::FromIterator, marker::PhantomData, sync::Arc, vec::Vec,
+};
+
+/// Loads a state snapshot repository from existing shards directory with state files.
+pub struct StateSnapshotRepositoryLoader<FileIo, State, HashType> {
+	file_io: Arc<FileIo>,
+	phantom_data: PhantomData<(State, HashType)>,
+}
+
+impl<FileIo, State, HashType> StateSnapshotRepositoryLoader<FileIo, State, HashType>
+where
+	FileIo: StateFileIo<StateType = State, HashType = HashType>,
+	HashType: Copy + Eq + Debug,
+{
+	pub fn new(file_io: Arc<FileIo>) -> Self {
+		StateSnapshotRepositoryLoader { file_io, phantom_data: Default::default() }
+	}
+
+	/// Load a state snapshot repository from an existing set of files and directories.
+	pub fn load_snapshot_repository(
+		&self,
+		snapshot_history_cache_size: usize,
+	) -> Result<StateSnapshotRepository<FileIo, State, HashType>> {
+		let snapshot_history = self.load_state_snapshot_history()?;
+
+		StateSnapshotRepository::new(
+			self.file_io.clone(),
+			snapshot_history_cache_size,
+			snapshot_history,
+		)
+	}
+
+	fn load_state_snapshot_history(&self) -> Result<SnapshotHistory<HashType>> {
+		let mut repository = SnapshotHistory::new();
+
+		let shards = self.file_io.list_shards()?;
+		debug!("Found {} shard(s) to load state from", shards.len());
+
+		for shard in shards {
+			let mut state_ids = self.file_io.list_state_ids_for_shard(&shard)?;
+			// Sort by id (which are timestamp), highest, i.e. newest, first
+			state_ids.sort_unstable();
+			state_ids.reverse();
+
+			let mut snapshot_metadata: Vec<_> = self.map_to_snapshot_metadata(&shard, state_ids);
+
+			if snapshot_metadata.is_empty() {
+				warn!(
+					"No (valid) states found for shard {:?}, initializing empty shard state",
+					shard
+				);
+				let initial_snapshot_metadata =
+					initialize_shard_with_snapshot(&shard, self.file_io.as_ref())?;
+				snapshot_metadata.push(initial_snapshot_metadata);
+			} else {
+				debug!(
+					"Found {} state snapshot(s) for shard {}, latest snapshot is {}",
+					snapshot_metadata.len(),
+					&shard,
+					snapshot_metadata.first().map(|f| f.state_id).unwrap_or_default()
+				);
+			}
+
+			let snapshot_history = VecDeque::from_iter(snapshot_metadata);
+
+			repository.insert(shard, snapshot_history);
+		}
+		Ok(repository)
+	}
+
+	fn map_to_snapshot_metadata(
+		&self,
+		shard: &ShardIdentifier,
+		state_ids: Vec<StateId>,
+	) -> Vec<StateSnapshotMetaData<HashType>> {
+		state_ids
+			.into_iter()
+			.flat_map(|state_id| {
+				self.file_io
+					.compute_hash(shard, state_id)
+					.map_err(|e| {
+						warn!(
+								"Failed to compute hash for state snapshot with id {}: {:?}, ignoring snapshot as a result",
+								state_id, e
+							);
+					})
+					.ok()
+					.map(|h| StateSnapshotMetaData::new(h, state_id))
```

Comment:
The error handling part will be more lines of code, however you put it. But I can use `match` instead of `map` and `map_err`, that does not matter much to me 👍